### PR TITLE
drag-n-drop upload plugin: rewrite

### PR DESCRIPTION
**Pull request by erykpiast**
## Implementation

Commented out code was reimplemented and battle-tested in existing DraftJS editor (prezly.com).

Public API was changed to improve separation of concerns and allow to configure more aspects of how the plugin works. Now it does only things it has to, and only these related with drag-n-drop and upload mechanism.

## Demo

No editable demo for now, sorry. @alfonsoperez, could you help with this?
Something to watch:  https://www.youtube.com/watch?v=fI5zXvLx-4Y

## Use-case

Sharing custom part of code (plugin configuration) between many different plugins. So you can use the same configuration function for plugin like "upload by button" or "upload images pasted with text".

### CODE REVIEW

**Comment by alfonsoperez** [comment]
@erykpiast what about providing a set of default functions for the common use case of uploading a file, like adding a placeholder in place and then replacing it with an image, it could be passed a custom styling.

I get the separation of concerns however but we could get both of things by using the default whenever customs are not specified.

Or at least have a self contained demo for the newbies like me that will provide examples of these functions, as all this is all pretty new still, which is what you mean right?, I could definitely craft a demo example, and when pushed to cdnjs if pushed I could do a jsfiddle version, what do you think.

**Comment by erykpiast** [comment]
@alfonsoperez I think it's must have!

**Comment by alfonsoperez** [comment]
Ok, I will take a stab at it over the weekend :+1:

**Comment by alfonsoperez** [comment]
By the way @erykpiast, let's see I have:
```
  handleUpload: (data: { files: Array<File>, formData: FormData }) =>
    Promise.all(data.files.map((file) => surveyUploadImages({ file }))),
```

How do you handle when `Promise.all` gets rejected?, like you would show a message in the placeholder or something?

**Comment by erykpiast** [comment]
Yeah, that's a `TODO` (look at comment in code).

**Comment by clwilliams** [comment]
Hi - I'm new to react / redux, and I've managed to get Draft JS up and running. I have been trying to get the upload images working as per the example on https://www.draft-js-plugins.com/plugin/image, yet I can't install the draft-js-drag-n-drop-upload-plugin. I've got the beta version of the draft js editor and so tried @beta too, but both times I get no valid versions available. Will this be fixed / made available do you know? I'm going around in circles with links to this post and https://github.com/draft-js-plugins/draft-js-plugins/issues/614. I don't understand why both issues closed if not working? Thanks

**Comment by alfonsoperez** [comment]
I was waiting for this to be merged to provide a helpful demo. I feel you!

File: draft-js-drag-n-drop-upload-plugin/src/handleDroppedFiles.js [hunk 1]
```diff
@@ -26,84 +28,61 @@ export default function onDropFile(config) {
...
+
+            if (!placeholderBlock) {
+              return editorState;
+            }
+
+            return handleBlock(editorState, placeholderBlock, file);
+          }, getEditorState());
+          setEditorState(editorStateWithImages);
+        }, (err) => {
+          console.error(err);
```
Comment (juliankrispel) [inline draft-js-drag-n-drop-upload-plugin/src/handleDroppedFiles.js]: It'd be appropriate to use a callback here, which you'd supply to the plugin config.

File: draft-js-drag-n-drop-upload-plugin/src/handleDroppedFiles.js [hunk 2]
```diff
@@ -26,84 +28,61 @@ export default function onDropFile(config) {
       // Set data {files: [Array of files], formData: FormData}
       const data = { files: [], formData };
       for (const key in files) { // eslint-disable-line no-restricted-syntax
-        if (files[key] && files[key] instanceof File) {
-          data.formData.append('files', files[key]);
-          data.files.push(files[key]);
+        const file = files[key];
```
Comment (juliankrispel) [inline draft-js-drag-n-drop-upload-plugin/src/handleDroppedFiles.js]: Getting a linting error here - `The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in` plz fix

File: draft-js-drag-n-drop-upload-plugin/src/utils/block.js [hunk 3]
```diff
@@ -1,8 +1,9 @@
+import { Entity } from 'draft-js';
```
Comment (nikgraf) [inline draft-js-drag-n-drop-upload-plugin/src/utils/block.js]: `Entity` is deprecated. Can you revert this change?

File: draft-js-drag-n-drop-upload-plugin/src/modifiers/modifyBlockData.js [hunk 4]
```diff
@@ -1,11 +1,11 @@
-import { EditorState } from 'draft-js';
+import { EditorState, Entity } from 'draft-js';
 
 export default function (editorState, key, data) {
   const currentContentState = editorState.getCurrentContent();
 
   const block = currentContentState.getBlockForKey(key);
   const entityKey = block.getEntityAt(0);
-  currentContentState.mergeEntityData(entityKey, data);
+  Entity.mergeData(entityKey, data);
```
Comment (juliankrispel) [inline draft-js-drag-n-drop-upload-plugin/src/modifiers/modifyBlockData.js]: `Entity` is deprecated, please revert

File: draft-js-drag-n-drop-upload-plugin/src/modifiers/replaceBlock.js [hunk 5]
```diff
@@ -10,14 +12,18 @@ export default function (editorState, blockKey, newType) {
...
+  const newState = EditorState.push(
+    editorState,
+    contentWithReplacedBlock,
+    'modify-block'
   );
 
-  // force to new selection
-  const newState = EditorState.push(editorState, content, 'modify-block');
+  // restore selection
   return EditorState.forceSelection(newState, editorState.getSelection());
```
Comment (juliankrispel) [inline draft-js-drag-n-drop-upload-plugin/src/modifiers/replaceBlock.js]: why use forceSelection here?

**Comment by juliankrispel** [comment]
@erykpiast hey dude, sorry that it took so long to get this reviewed but please if you're still interested in finishing it lmk either way

**Comment by erykpiast** [comment]
@juliankrispel I'll try this weekend. I'm no longer involved in project relying on DraftJS so I have to do it after hours, I hope you understand.

**Comment by juliankrispel** [comment]
@erykpiast dude of course! Thank you so much! If you need me to review stuff or have questions feel free to reach out to me on [the draft js slack](https://draftjs.herokuapp.com/)

**Comment by juliankrispel** [comment]
@erykpiast I'd be happy to take this over, could you reach out to me on the draft js slack for a convenient time?

**Comment by juliankrispel** [comment]
@erykpiast wohooooo

**Comment by erykpiast** [comment]
Hey, I've just merged the branch with upstream, I have to test if everything still works. I'm totally out of topic now (well, it's five months since I did the change), but I'll try at least.

**Comment by juliankrispel** [comment]
@erykpiast beautiful! Wanna pair on it?

**Comment by juliankrispel** [comment]
@erykpiast feel free to ping me on the draft js slack :+1:

**Comment by juliankrispel** [comment]
@erykpiast any chance I can help? :D

File: draft-js-drag-n-drop-plugin/src/createDecorator.js [hunk 6]
```diff
@@ -21,11 +21,12 @@ export default ({ store }) => (WrappedComponent) => (
     }
 
     render() {
-      const readOnly = store.getReadOnly();
+      // const readOnly = store.getReadOnly();
```
Comment (erykpiast) [inline draft-js-drag-n-drop-plugin/src/createDecorator.js]: I've no idea why I've commented this out.

**Comment by erykpiast** [comment]
Sure. If you have a project where you can plug this in and check if it works for you. We'd define some test cases together.

**Comment by kopacki** [comment]
@erykpiast @juliankrispel 
Hello gents! Can I help with this one? It looks like the one I would need soon in my project. I'd be glad if I could help with merging this PR in.

**Comment by erykpiast** [comment]
Hey @kopacki, I guess broken/missing tests prevent this one to be merged. I'm no longer involved in DraftJS development (this PR is 6 months-old, a lot of happened during this period) so feel free to take this.

**Comment by idoromm** [comment]
Any update on this? I think it's an extremely must-have feature for Draft JS...

**Comment by juliankrispel** [comment]
Yea it’s a bigger piece of work and I‘m not sure I have room in my head or
on my schedule to tackle it soon. At least not alone. Anyone who fancies
being a co-maintainer wants to tackle this with me? Cc @tumetus? 🙃❤️
